packages/financial-templates-lib/src/clients/OptimisticOracleClient.ts

```diff
@@ -154,32 +156,56 @@ export class OptimisticOracleClient {
 
   public async update(): Promise<void> {
     // Determine earliest block to query events for based on lookback window:
+    const netId = await this.web3.eth.getChainId();
     const [averageBlockTime, currentBlock] = await Promise.all([
-      averageBlockTimeSeconds(),
+      averageBlockTimeSeconds(netId),
       this.web3.eth.getBlock("latest"),
     ]);
     const lookbackBlocks = Math.ceil(this.lookback / averageBlockTime);
     const earliestBlockToQuery = Math.max(currentBlock.number - lookbackBlocks, 0);
 
-    // Fetch contract state variables in parallel.
-    // Note: We can treat all events by default as normal OptimisticOracle events because in most cases we only
-    // need to read the requester, identifier, timestamp, and ancillary data which are emitted in both Skinny
-    // and normal OptimisticOracle events.
-    const [requestEvents, proposalEvents, disputeEvents, settleEvents, currentTime] = await Promise.all([
-      (this.oracle.getPastEvents("RequestPrice", { fromBlock: earliestBlockToQuery }) as unknown) as Promise<
-        RequestPrice[] | SkinnyRequestPrice[]
-      >,
-      (this.oracle.getPastEvents("ProposePrice", { fromBlock: earliestBlockToQuery }) as unknown) as Promise<
-        ProposePrice[] | SkinnyProposePrice[]
-      >,
-      (this.oracle.getPastEvents("DisputePrice", { fromBlock: earliestBlockToQuery }) as unknown) as Promise<
-        DisputePrice[] | SkinnyDisputePrice[]
-      >,
-      (this.oracle.getPastEvents("Settle", { fromBlock: earliestBlockToQuery }) as unknown) as Promise<
-        Settle[] | SkinnySettle[]
-      >,
-      this.oracle.methods.getCurrentTime().call(),
-    ]);
+    // If max blocks to search is defined, send multiple web3 requests, otherwise search all block history in one search.
+    let blockSearchConfig = { fromBlock: earliestBlockToQuery, toBlock: currentBlock.number };
+    if (this.blocksPerEventSearch !== null) {
+      blockSearchConfig.toBlock = Number(earliestBlockToQuery) + Number(this.blocksPerEventSearch);
+    }
+
+    const currentTime = await this.oracle.methods.getCurrentTime().call();
+    let requestEvents: (RequestPrice | SkinnyRequestPrice)[] = [];
+    let proposalEvents: (ProposePrice | SkinnyProposePrice)[] = [];
+    let disputeEvents: (DisputePrice | SkinnyDisputePrice)[] = [];
+    let settleEvents: (Settle | SkinnySettle)[] = [];
+
+    while (blockSearchConfig.fromBlock <= currentBlock.number) {
+      this.logger.debug({
+        at: "OptimisticOracleClient",
+        message: "Fetching events with block search config",
+        blockSearchConfig,
+      });
+      const eventSearchResults = await Promise.all([
```

Comment:
this is not parallel, no? would it be better to gather up all the promises and then fire them at once? the counter-argument to doing this would be that this will place a lot more load onto the PRC provider.

Reply:
Oh good call I'll parallelize it and test in prod

Reply:
excellent, thanks!